File: Algorithm/BOJ/C++/Dynamic Programming/14501.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


int solution(vector<vector<int>> time_reward_lists) {

	vector<int> answer;	
	for (const auto& element: time_reward_lists) {
		answer.push_back(element[1]);
	}
	answer.push_back(0);

	for (int idx = time_reward_lists.size() - 1; idx >= 0; idx--) {
		if (time_reward_lists[idx][0] + idx > time_reward_lists.size()) {
			answer[idx] = answer[idx + 1];
		} 
		else {
			answer[idx] = max(answer[idx + 1], time_reward_lists[idx][1] + answer[idx + time_reward_lists[idx][0]]);
		}
	}

	return answer[0];
}


int main() {
	int n, input;
	vector<vector<int>> time_reward_lists;

	cin >> n;
	for (int i = 0; i < n; i++) {
		vector<int> element(2);
		for (auto& input : element) {
			cin >> input;
		}
		time_reward_lists.push_back(element);
	}

	cout << solution(time_reward_lists);

	return 0;
}

File: Algorithm/BOJ/Python/Dynamic Programming/14501.py
def solution(time_reward_lists):

    answer = list(map(lambda x: x[1], time_reward_lists))
    answer.append(0)


    for idx in range(len(time_reward_lists) - 1, -1, -1):    
        if time_reward_lists[idx][0] + idx > len(time_reward_lists):
            answer[idx] = answer[idx + 1]
        else:
            answer[idx] = max(answer[idx + 1], time_reward_lists[idx][1] + answer[idx + time_reward_lists[idx][0]])
                
    return print(answer[0])
        

n = int(input())
time_reward_lists = []
for _ in range(n):
    input_ = list(map(int, input().split()))
    time_reward_lists.append(input_)

solution(time_reward_lists)
